task/download/download.go

```diff
@@ -44,40 +45,66 @@ type downloader struct {
 // whether it should be re-cloned.
 func getHashOfRepo(repo *task.Repository) string {
 	data := fmt.Sprintf("%s|%s|%s|%s", repo.Clone, repo.Ref, repo.Sha, repo.Download)
+	return getHash(data)
+}
+
+func getHash(s string) string {
 	hash := sha256.New()
-	hash.Write([]byte(data))
+	hash.Write([]byte(s))
 	return hex.EncodeToString(hash.Sum(nil))
 }
 
-func (d *downloader) Download(ctx context.Context, repo *task.Repository) (string, error) {
-	if repo == nil {
-		return "", errors.New("no repository provided to download")
+func (d *downloader) Download(ctx context.Context, repo *task.Repository, urls *task.ExecutableUrls) (string, error) {
+	if urls != nil {
+		return d.handleDownloadExecutable(ctx, urls)
+	} else if repo != nil {
+		return d.handleDownloadRepo(ctx, repo)
 	}
-	log := logger.FromContext(ctx)
+	return "", errors.New("no repository or executable urls provided to download")
+}
+
+func (d *downloader) handleDownloadExecutable(ctx context.Context, urls *task.ExecutableUrls) (string, error) {
+	osWithArch := fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH)
+	url, ok := (*urls)[osWithArch]
+	if !ok {
+		return "", fmt.Errorf("os and architecture [%s] is not specified in task.yml file", osWithArch)
+	}
+
+	dest := filepath.Join(d.getBaseDownloadDir(), getHash(url))
```

Comment:
dest of the executable better can be 

getBaseDownloadDir() +"/" + typeOfExecutable 

For Runner, we don't need to keep multiple downloaded executables of the same type, because then runner will run out of space, and incident will happen.

Delegate has been doing this way ("replace whenever download") for a long time.

Reply:
Thanks for the comment! Done - The final download folder will be `{baseDir}/{taskType}/{version}`, so that it is a cache hit when searching for an executable of a given task type and version. 

If it is not a cache hit, we are deleting `{baseDir}/{taskType}` before downloading the binary, so that we get rid of any already downloaded files for the same task type.

Reply:
sounds good ! thanks